File: examples/connext_dds/request_reply/cs/PrimesRequester.cs
/*
* (c) Copyright, Real-Time Innovations, 2022.  All rights reserved.
* RTI grants Licensee a license to use, modify, compile, and create derivative
* works of the software solely for use with RTI Connext DDS. Licensee may
* redistribute copies of the software provided that all such copies are subject
* to this license. The software is provided "as is", with no warranty of any
* type, including any warranty for fitness for any purpose. RTI is under no
* obligation to maintain or support the software. RTI shall not be liable for
* any incidental or consequential damages arising out of the use or inability
* to use the software.
*/

using System;
using System.Threading;
using System.Threading.Tasks;
using Rti.Dds.Core;
using Rti.Dds.Domain;
using Rti.RequestReply;
using Rti.Dds.Subscription;

namespace PrimeCalculator
{
    /// <summary>
    /// Example application that publishes PrimeNumberReply
    /// </summary>
    public sealed class PrimesRequester : IApplication
    {
        private const string qosProfileName =
            "RequestReplyExampleProfiles::RequesterExampleProfile";
        private readonly DomainParticipant participant;
        private readonly Requester<PrimeNumberRequest, PrimeNumberReply> requester;
        private readonly int n;
        private readonly int primesPerReply;

        /// <summary>
        /// Creates a DomainParticipant, Topic, Requester and DataWriter<PrimeNumberReply>.
        /// </summary>
        public PrimesRequester(int domainId, int n, int primesPerReply)
        {
            this.n = n;
            this.primesPerReply = primesPerReply;

            // The requester is created within a DomainParticipant
            participant = DomainParticipantFactory.Instance
                .CreateParticipant(domainId);

            var qosProvider = QosProvider.Default;

            // Create a Requester with a service name, and a QoS profile
            //
            // BuildRequester is a DomainParticipant extension method included
            // in the namespace Rti.Dds.RequestReply. It returns a builder
            // that allows configuring the arguments to create the Requester.
            requester = participant.BuildRequester<PrimeNumberRequest, PrimeNumberReply>()
                .WithServiceName("PrimeCalculator")
                .WithDataWriterQos(qosProvider.GetDataWriterQos(qosProfileName))
                .WithDataReaderQos(qosProvider.GetDataReaderQos(qosProfileName))
                .Create();
        }

        /// <summary>
        /// Sends the request, waits for the replies asynchronously, and prints
        /// the results.
        /// </summary>
        public async Task Run(CancellationToken cancellationToken)
        {
            // Wait for at least one replier
            while (requester.GetMatchedReplierCount() == 0)
            {
                await Task.Delay(millisecondsDelay: 100, cancellationToken);
            }

            requester.SendRequest(new PrimeNumberRequest(n, primesPerReply));

            bool inProgress = true;
            while (inProgress)
            {
                await requester.WaitForRepliesAsync(1, cancellationToken);

                using LoanedSamples<PrimeNumberReply> replies =
                    requester.TakeReplies();

                // Print the prime numbers we receive
                foreach (var reply in replies.ValidData())
                {
                    foreach (var prime in reply.primes)
                    {
                        Console.Write($"{prime} ");
                    }

                    if (reply.status != PrimeNumberCalculationStatus.REPLY_IN_PROGRESS)
                    {
                        inProgress = false;
                        if (reply.status == PrimeNumberCalculationStatus.REPLY_ERROR)
                        {
                            Console.Write("Replier returned an error");
                        }
                        else
                        {
                            Console.Write("DONE");
                        }
                    }

                    Console.WriteLine();
                }
            }
        }

        /// <summary>
        /// Disposes the requester and the participant.
        /// </summary>
        public void Dispose()
        {
            requester.Dispose();
            participant.Dispose();
        }
    }
} // namespace PrimeCalculator

File: examples/connext_dds/request_reply/py/primes_requester.py
#
# (c) 2023 Copyright, Real-Time Innovations, Inc.  All rights reserved.
#
# RTI grants Licensee a license to use, modify, compile, and create derivative
# works of the Software solely for use with RTI products.  The Software is
# provided "as is", with no warranty of any type, including any warranty for
# fitness for any purpose. RTI is under no obligation to maintain or support
# the Software.  RTI shall not be liable for any incidental or consequential
# damages arising out of the use or inability to use the software.
#

import argparse
import time
from typing import Sequence

import rti.connextdds as dds
from rti.request import Requester
import Primes


def requester_main(domain_id, n, primes_per_reply):
    # Get the QoS from a profile in USER_QOS_PROFILES.xml (the default
    # QosProvider will load the USER_QOS_PROFILES.xml file from the current
    # working directory)
    qos_provider = dds.QosProvider.default
    writer_qos = qos_provider.datawriter_qos_from_profile(
        "RequestReplyExampleProfiles::RequesterExampleProfile"
    )
    reader_qos = qos_provider.datareader_qos_from_profile(
        "RequestReplyExampleProfiles::RequesterExampleProfile"
    )

    participant = dds.DomainParticipant(domain_id)

    requester = Requester(
        request_type=Primes.PrimeNumberRequest,
        reply_type=Primes.PrimeNumberReply,
        participant=participant,
        service_name="PrimeCalculator",
        datawriter_qos=writer_qos,
        datareader_qos=reader_qos,
    )

    print(f"Waiting to discover replier on domain {domain_id}")

    while requester.matched_replier_count == 0:
        time.sleep(0.1)

    prime_number_request = Primes.PrimeNumberRequest(
        n=n, primes_per_reply=primes_per_reply
    )

    print(
        f"Sending a request to calculate the prime numbers <= {n} in sequences of {primes_per_reply} or fewer elements"
    )

    request_id = requester.send_request(prime_number_request)

    max_wait = dds.Duration.from_seconds(20)
    in_progress = True
    while in_progress:
        if not requester.wait_for_replies(
            max_wait, related_request_id=request_id
        ):
            raise dds.TimeoutError("Timed out waiting for replies")

        # It's possible to call requester.take_replies(), but that method also
        # returns SampleInfo objects, which we don't need here.
        replies: Sequence[
            Primes.PrimeNumberReply
        ] = requester.reply_datareader.take_data()
        for reply in replies:
            print(list(reply.primes))

            if (
                reply.status
                != Primes.PrimeNumberCalculationStatus.REPLY_IN_PROGRESS
            ):
                in_progress = False
                if (
                    reply.status
                    == Primes.PrimeNumberCalculationStatus.REPLY_ERROR
                ):
                    raise RuntimeError("Error in Replier")

    print("DONE")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="RTI Connext DDS Example: Request-Reply (Requester)"
    )
    parser.add_argument(
        "-d",
        "--domain",
        type=int,
        default=0,
        help="DDS Domain ID (default: 0)",
    )
    parser.add_argument(
        "-p",
        "--primes-per-reply",
        type=int,
        default=5,
        help="Number of primes per reply (min: 5, default: 5)",
    )
    parser.add_argument(
        "n", type=int, help="Request will retrieve prime numbers <= n"
    )

    args = parser.parse_args()
    assert 0 <= args.domain < 233
    assert args.primes_per_reply >= 5

    requester_main(args.domain, args.n, args.primes_per_reply)
